dopamine/discrete_domains/run_experiment.py

```diff
@@ -570,64 +601,59 @@ def __init__(
         environment, and returns an agent.
       create_environment_fn: A function which receives a problem name and
         creates a Gym environment for that problem (e.g. an Atari 2600 game).
-      max_simultaneous_iterations: int, maximum number of iterations running
+      num_simultaneous_iterations: int, number of iterations running
         simultaneously in separate threads.
       **kwargs: Additional positional arguments.
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
-    self._eval_period = max_simultaneous_iterations
+    self._eval_period = num_simultaneous_iterations
+    self._num_simultaneous_iterations = num_simultaneous_iterations
     self._output_lock = threading.Lock()
-    self._experience_queue = queue.Queue(max_simultaneous_iterations)
-    self._training_queue = queue.Queue(max_simultaneous_iterations)
+    self._training_queue = queue.Queue(num_simultaneous_iterations)
 
     super(AsyncRunner, self).__init__(
         base_dir=base_dir, create_agent_fn=create_agent_fn,
         create_environment_fn=create_environment_fn, **kwargs)
 
-  # TODO(aarg): Decouple experience generation from training.
   def _run_iterations(self):
     """Runs required number of training iterations sequentially.
 
     Statistics from each iteration are logged and exported for tensorboard.
 
     Iterations are run in multiple threads simultaneously (number of
-    simultaneous threads is specified by `max_simultaneous_iterations`). Each
+    simultaneous threads is specified by `num_simultaneous_iterations`). Each
     time an iteration completes a new one starts until the right number of
     iterations is run.
     """
-    # TODO(aarg): Change the thread management to an implementation with queues
-    # and with a fix number of thread workers. With the revamp, training might
-    # go inline to this method.
-    def _start_iteration(*args):
-      thread = threading.Thread(target=self._run_one_iteration, args=args)
-      thread.start()
-      return thread
-
-    training_worker = threading.Thread(target=self._run_training_steps)
-    training_worker.start()
-    experience_threads = []
+    experience_queue = queue.Queue()
+    worker_threads = []
+
+    for _ in range(self._num_simultaneous_iterations):
```

Comment:
There's an awkwardness and complexity that comes from the different scopes of each task.  "experience-gathering" tasks are to run an entire iteration, while "training step" tasks are for just a single step.  An experience-gathering task does so much more than gather experience (and, further, when it's in eval mode, doesn't even gather experience).

It *feels* like there should be an approach where each task is of the same lifetime.  And that if the code followed that approach then it'd be easier to keep track of what happens when (and it might even use fewer instance variables).

Do you have any ideas?

Reply:
If the implementation was easy, what would the ideal scope of a task be?

- iteration level: Since each iteration (at least so far as **either** environment-interaction **or** training is concerned) is single-threaded, it's easy to think about tasks being at this level.  The tasks themselves, are somewhat complicated but not so bad.  Given the current implementation, there would need to be some coordination around result summaries.

- step level: If iterations are at the step level, then workers can be so dumb.  This is attractive.  But then who's in charge of managing the simultaneous iterations?  Those could be tracked in a single thread.  I can imagine a system where that management is clear and makes sense.  But that system would take work to write.

----------------
I think I like iteration-level tasks, then?  But then how to untie training and environment-interaction?

Mmm.  Could you remind me what we get from decoupling them?

Reply:
Discussed the benefits of decoupling in person, adding the answer here for the record:
```
"What are the benefits of having them decoupled?" it runs faster. Originally these tasks are sequential, so during an environment step, the computing units are idle. With decoupling, training can run while waiting for the environment's response. This also allows to run more training steps without slowing down training (wanted to make this possible in a next patch)
```

Regarding the level of the tasks:
- it seems indeed that iteration level is the right choice. Otherwise I don't see an easy way to manage iteration length, running statistics, managing eval iterations, etc. It seems to add a lot of complexity.

The way i see this would work is really not nice: each time a low-level task is completed, we have a bunch of condition that decide which is the next task that needs to be added to the task queue, and which metrics needs to be aggregated.

If we choose iteration level, then we still need to trigger training step task at the step level because we don't want to wait for an iteration to finish to start training.

So I like your idea but I'm not too sure how that would work nicely. Do you know?

Reply:
Trying to define the desired rules for execution order within an iteration:

* For every env-step *i* there's a training-step *i*
* There's an *n* such that no env step *i+n* can happen before training step *i* completes
* There's an *m* such that no training step *i+m* can happen before env step *i* completes
* No more than 1 env step should be executed at the same time
* No more than 1 training step should be executed at the same time

Reply:
The code as-written shares its training worker pool across all iterations.  I'm going to propose here that this behavior is neither necessary nor desirable.

Reply:
With that freedom, I think the following works.  I'll just talk about training iterations, and I think eval iterations will follow naturally.

* Master thread:
  * creates an iteration queue
  * enqueues `num_iterations - start_iterations` copies of `(self._run_one_iteration, (iteration, False))` tuples
  * creates a shared `n_completed` iteration counter
  * spawns `n_simultaneous_iterations` iteration workers threads with thread-local `_environment`s to work the queue
  * joins and terminates iteration workers

* An iteration worker:
  * resets its thread-local environment
  * creates a thread-local experience queue and a thread-local training queue
  * creates step counters for experience and training that are shared with its child threads
  * enqueues `max_steps_per_episode` jobs into the experience queue and into the training queue
  * spawns 1 experience worker and 1 training worker
  * joins each queue
  * writes summaries (with a global lock)
  * increments the experience counter (with the same global lock)

* An experience worker examines the experience queue and:
  * kills itself if asked by its iteration worker
  * checks if the environment is in a terminal state.  If so, returns immediately.
  * `while (experience_step > training_step + n):`
    * `sleep 1` (or whatever)
  * runs a step
  * increments the experience counter
* Analogously, the training counter:
  * kills itself if asked by its iteration worker
  * checks if the environment is in a terminal state.  If so, returns immediately.
  * `while (training_step > experience_step + m):`
    * `sleep 1`
  * runs a training step
  * increments the training counter

Reply:
This respects all the above rules of intra-iteration execution order.  I also think it encapsulates threads and variables within their minimal scope.  With one explicit (and differentiated) experience worker and training worker per iteration, I also think that an implementation based around this design will be easier to think about.

The only gross part I notice offhand is that experience and training worker for iteration *i* need access to *i*'s thread-local variables (or, alternatively, that iteration *i* needs access to its workers' thread-local variables).  I think this is possible given the thread-local library you wrote, but will also be a little hacky.

How's it look to you?

Reply:
To answer you first comment:
```There's an m such that no training step i+m can happen before env step i completes```
that's true for all m
```No more than 1 env step should be executed at the same time``` no I think many env steps can be executed at the same time
```No more than 1 training step should be executed at the same time```
-> several steps can be run at the same time (typically `num_iterations` at the same time)

Reply:
> No more than 1 env step should be executed at the same time
>> no I think many env steps can be executed at the same time
You may have read past the context for this to be *within* an iteration.  Unless environments are stateless, I don't know what it would mean for 2 env steps to be run concurrently in 1 environment.

Reply:
Makes sense, thank you for the clarification. I think at least, we could hand-off the queue-based training step management to `_run_one_iteration` with a queue size of 1. I like this idea, and this can be done while keeping experience generation in the current thread.
That means the training queue could be initialized in `_run_one_training_phase` and passed to the called methods so it gets filled during experience generation.

Reply:
There is also the following alternative:

Let N be the number of experience workers

- Start N iteration threads
- Start N environment threads
- enqueue X iterations tasks
- within each iteration thread:
* pop an iteration task
* add `min_steps` to the `training_environment_steps` queue
* wait for `min_steps` to be completed in the statistics queue
* pop the statistics for the first `min_steps` in the statistics queue
* logs the statistics, computes metrics and complete all the exports
- within each environment thread, run the following task:
* pop a step in the `training_environment_steps` queue
* stats, done = _run_step()
* add a training step to the training step queue
* if done add (num steps in episodes, stats) to the statistics queue
* else start again


Each of these approach is a large revamp and is not required for the software to work efficiently. I added it as a todo, we can see how to prioritize this change.